src/mscorlib/corefx/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
```

Comment:
> Pinning for synchronous PInvokes is close to free, it is much cheaper than overhead of safe handles that the current scheme is based on.

So pinning a char/byte[] array is close to free? That surprises me. I would have expected it to create significant grief for the GC when collecting.

Where is the overhead for SafeHandle? Just that they're registered? Or is it through interop? I have no particular affinity for using SafeHandle in the P/Invoke signatures if that makes a difference. With this design we aren't really creating a lot, but for internal usage there is no _real_ need to use SafeHandle. The intent here is not to be passing these Buffers around- they're meant to be used within a using block.